### ISSUE TITLE
Promise support

### ISSUE THREAD

**Issue by tbranyen** [juliangarnier/anime#4]
Would it be out-of-scope to provide a Promise for animation completion? Something like Web Animations `finished` Promise:

``` javascript
var myAnimation = anime({
  targets: ['.blue', '.green'],
  translateX: '13rem',
  rotate: 180,
  borderRadius: 8,
  duration: 2000,
  loop: true
});

myAnimation.finished.then(() => {
  /* ... */
});
```

**Comment by tbranyen** [comment]
Since `loop: true` indicates a Promise that never really ends, this might be too complex to simply tack on. I think Web Animations re-assign the finished property, which seems rather ugly for this.

**Comment by tbranyen** [comment]
Maybe something like this (inspired by jQuery) could work:

``` js
var myAnimation = anime({
  targets: ['.blue', '.green'],
  translateX: '13rem',
  rotate: 180,
  borderRadius: 8,
  duration: 2000,
  loop: true
});

// Wait for the current Promise to complete. Is called recursively
// upon the prior loop's completion.
function animationCompleted() {
  return myAnimation.promise().then(() => {
    console.log('Animation loop finished');
    return animationCompleted();
  });
}

// Kick off infinite-Promise resolution.
animationCompleted();
```

The `promise()` invocation would infer a new Promise that will resolve when the current animation completes. If `loop: false` is set, then the same Promise would be returned.

**Comment by james-brndwgn** [comment]
Perhaps I'm understanding you incorrectly, but wouldn't you require a callback? A promise would imply that you want to asynchronously run an animation, whereas really you're looking for when the animation is completed, thus a finished callback would be more appropriate.

**Comment by james-brndwgn** [comment]
Ah, I see. I did misunderstand. I assume, you're looking to chain animations

**Comment by joseluisq** [comment]
:+1:

**Comment by juliangarnier** [comment]
I like the idea, and wanted to add promise initially, but I never really had the need to use it at this point.

But it's on my list.

**Comment by tbranyen** [comment]
@juliangarnier This issue was opened to get your blessing for a PR. If I implement it will you consider?

**Comment by luknei** [comment]
It is very useful when you have separate animations that have to be put into a sequence. @tbranyen I'd like to contribute.

**Comment by juliangarnier** [comment]
@tbranyen Yes I will!

**Comment by juliangarnier** [comment]
I just created a branch for all the new developments that will potentially be released in v1.1.
Please PR here from now https://github.com/juliangarnier/anime/tree/v1.1

Thanks

**Comment by tbranyen** [comment]
Cool @juliangarnier hoping to tackle this after work today, I'll keep you posted.

**Comment by joseluisq** [comment]
> I just created a branch for all the new developments that will potentially be released in v1.1.
> Please PR here from now https://github.com/juliangarnier/anime/tree/v1.1
> 
> Thanks

@juliangarnier cool 👍

**Comment by devmario** [comment]
@tbranyen If you implement this issue, please let me know.
We seem to be better when implemented in parallel.
This is my PR #23

**Comment by SaulDoesCode** [comment]
There is a new Proposed syntax for promise based animation event detection in the latest branches of anime @juliangarnier has implemented some in a branch of his and I have implemented an event system based version that has the exact same syntax on [my fork ](https://github.com/SaulDoesCode/anime).

The proposed syntax looks like this
```javascript
    var i = 0;
    myAnimation.on('update', function(anim) {
           console.log('update step :',i++);     // 1,2,3,4,5...
    });

    myAnimation.once('complete').then(function(anim) {
      console.log('completed once');
      anim.play(); // will re-play the animation only once
    });
   
   // basic layout for reacting on an event
  animation.on( eventname , callback);
  animation.once( eventname ) // -> promise
```

**Comment by tbranyen** [comment]
:-1: Sorry that looks very confusing. I see events, callbacks, and Promises control flow all rolled into the exact same API.

Why was the `promise()` method not sufficient? What does this new API solve?

**Comment by SaulDoesCode** [comment]
``.promise()`` leaves out room for different events.
Do you have a better proposed syntax , I suggest perhaps adding
```javascript
  animation.complete.then( func ):
  animation.begin.then( func );
  animation.pause.then( func );
```

**Comment by tbranyen** [comment]
I see, the goal of the Promise is to represent an animation transaction, it is not used for general events. My recommendation would be to separate the two entirely.

You have an event based system to hook into the animation states. And a very simple Promise API to know when the animation completes. If `loop: true` is set, then calling `promise()` again gets a new Promise since a new animation has started.

**Comment by SaulDoesCode** [comment]
Will the purpose of ``.promise()`` be solely for when the animation finishes? 
```javascript  
 // I think then that this promise getter is simpler than .promise()
 // it has more symantic value than .promise() as the user knows what the promise 
// is for
 animation.complete // -> promise
```

**Comment by tbranyen** [comment]
I'm afraid I disagree, providing events like `update` with a Promise is confusing (since Promise resolution is immutable once set). If you ignore that event, the API becomes inconsistent, which is less than ideal... With the API you have above I can't tell if the `begin` callback can get called multiple times, while the promise only gets resolved for the first. Can you tell just by looking at it?

It makes significantly more sense to model the Promise after the animation transaction. They are parallels in concept and implementation. With the suggestion you've made to use `complete` there would be no way to hook into various `completes` with `loop: true`, right?

**Comment by SaulDoesCode** [comment]
You're not meant to use update with a promise you can if you have to but it will run once as you say. That is why .on is provided promises in the case of the anime-next fork is optional you can do all the things with out it, promises give those who want to use them choice , benefits and all the problems that come with them.

**Comment by SaulDoesCode** [comment]
In anime-next I implemented an interloop event  since complete would not be viable as it looks instead optionally you could react to when a loop ends and starts

**Comment by SaulDoesCode** [comment]
All the promise events automatically just run once this is how it is implemented, maybe read  The source code of @juliangarnier's promise branch and read anime-next's code before making certain assuptions

**Comment by rohan-deshpande** [comment]
I've built a promise based animation class. Works really well. Currently I've only got it working with "tweens" not looping animations but I'm working on it. 

Anyway the basic premise is that you start the animation and return a promise from that method, then resolve the promise with an `end` method. In my class the end method is private. There's a public `stop` method that can be called which also returns the promise then calls end. 

Anyway it's really useful for chaining. Looks like right now there's no way of chaining with animejs? That's a real shame.

**Comment by rohan-deshpande** [comment]
Okay so here we go, I built a Promise wrapper for `animejs` http://codepen.io/rohandeshpande/pen/PzJdkz

It uses Bluebird and `Promise.defer()` but if that bothers you then it's easy to change I think.

**Comment by exah** [comment]
Hi! I created simple wrapper around Anime that resolves Promise when animation is complete for my own needs (see [Why?](https://github.com/exah/promise-anime#why) section). It doesn't include Anime as dependency so it will work with future versions while there is `complete` parameter and anime in your project. 

[Example](http://codepen.io/exah/pen/OWOaPG?editors=0110)

**Comment by tbranyen** [comment]
Implementation here:

https://github.com/juliangarnier/anime/pull/106#issuecomment-276546443

**Comment by juliangarnier** [comment]
AND merged.

**Comment by franzanzanz** [comment]
Don't know if it's a known issue but I thought I'd note it here as I just stumbled upon this. Under certain circumstances the promise is resolved multiple times:

If I start a run-once animation and reset it before it is completed – via restart() pause() or reset() – then start it again, letting it finish, the promise will be resolved **_twice_** whereas the 'complete' callback only fires once (as expected). This behaviour scales with the times you cancel the animation: let's say I cancel the animation 5 times and let it complete on the 6th run the promise will resolve 6 times.

If you have a use case like this use the 'complete' callback instead to avoid unintended function executions.
